Added Aim callback handler

tmynn opened the pull request:
These changes include:
- Added an Aim callback handler: `AimCallback` to track LLM-given prompts and outputs
- Provided a demo usage script of `AimCallback` with `GPTListIndex`

![llamaindex-demo-min](https://user-images.githubusercontent.com/23078323/236964969-66e8a017-eb5c-4d25-9310-5618b22abf64.gif)

Disiok commented:
Sweet! Taking a look!

Disiok commented:
> These changes include:
> 
> * Added an Aim callback handler: `AimCallback` to track LLM-given prompts and outputs
> * Provided a demo usage script of `AimCallback` with `GPTListIndex`
> 
> ![llamaindex-demo-min](https://user-images.githubusercontent.com/23078323/236964969-66e8a017-eb5c-4d25-9310-5618b22abf64.gif) [ ![llamaindex-demo-min](https://user-images.githubusercontent.com/23078323/236964969-66e8a017-eb5c-4d25-9310-5618b22abf64.gif) ](https://user-images.githubusercontent.com/23078323/236964969-66e8a017-eb5c-4d25-9310-5618b22abf64.gif) [ ](https://user-images.githubusercontent.com/23078323/236964969-66e8a017-eb5c-4d25-9310-5618b22abf64.gif)

Curious, are request and responses automatically matched together with event ID in the UI?

Disiok commented on llama_index/callbacks/aim.py:
```diff
@@ -0,0 +1,140 @@
+        self._run = None
+        self._run_hash = None
+
+        self._llm_response_step = 0
```
curious: what is this for?

tmynn replied:
it is to track the steps of LLM-type events and provide the `step` argument of `Run.track` method.

Disiok replied:
conceptually, what does step mean? Is it the number of LLM call in a sequence of calls in a since session?

Disiok replied:
related to this, if the LLM calls are async, how should we think about it?

Disiok commented on llama_index/callbacks/aim.py:
```diff
@@ -0,0 +1,140 @@
+            payload (Optional[Dict[str, Any]]): payload to store.
+            event_id (str): event id to store.
+        """
+        if event_type is CBEventType.LLM and payload:
```
curious: Can Aim handle non-LLM call events? We also track e.g. text chunking, embedding calls.

tmynn replied:
Sure it can. Thanks for the suggestion. 🚀
I will add chunking and embedding calls. Do you think any other type of event can be useful to track?

logan-markewich commented on llama_index/callbacks/aim.py:
```diff
@@ -0,0 +1,140 @@
+
+            self._llm_response_step += 1
+
+    def on_event_end(
```
Not sure if you want track event runtime at all, but this would be the place to do it in the future

logan-markewich replied:
Also, I thought the prompt/response got logged on event end? 👀

logan-markewich commented:
Looks super cool! Very cool to visualize how the output changes across multiple LLM calls (which also helps debug issues with the refine process as well! 💪)

logan-markewich commented:
Let us know if you need a hand with the linting/tests 👍

logan-markewich commented on llama_index/callbacks/aim.py:
```diff
@@ -0,0 +1,140 @@
+logger = logging.getLogger("root").setLevel(logging.WARNING)
+
+
+class AimCallback(BaseCallbackHandler):
```
Actually, I changed my mind hah

tmynn commented:
> Curious, are request and responses automatically matched together with event ID in the UI?

Hey @Disiok thanks a lot for the comments.☺️
It is actually matched with `step`, and for each step/event we log both the prompt and output.

tmynn commented on llama_index/callbacks/aim.py:
```diff
@@ -0,0 +1,140 @@
+from aim.ext.system_info import DEFAULT_SYSTEM_TRACKING_INT
+import logging
+
+logger = logging.getLogger("root").setLevel(logging.WARNING)
```
Hi, @logan-markewich thanks a lot for the comments ☺️.
Just changed it to `__name__`.

logan-markewich commented:
@tmynn some minor changes to where we call the callbacks just changed: https://github.com/jerryjliu/llama_index/pull/2669

Also a small bug fix where event_end/start were mixed up, which will change where you log the event data (should be on event end I think) https://github.com/jerryjliu/llama_index/pull/3146

Are you working on further changes/features? We can help finish things off if not 👍🏻

tmynn commented:
@logan-markewich Thanks for the updates.
The latest push contains:
- Added chunk tracking
- Moved to track to `on_event_end` method
- Added `run_params` argument to provide run arguments as a dict

Failed to capture an event of type `CBEventType.EMBEDDING`.
Please let me know what else can be tracked. 

Would greatly appreciate help with linting/tests. 🙌🏻

logan-markewich commented:
@tmynn awesome! I just cleaned up the linting/build errors

For the embedding tracking, it really only tracks the start/end of each occurrence (and so AIM could track stats related to embedding calls possibly). Each start/end event would have a matching ID, so that the duration can be counted.

Not sure if this fits into AIM's scope. Can also be left for a future PR as well

logan-markewich commented:
Ah, I notice a bug or two. Going to clean that up and maybe write a docs page, should have time to have it ready sometime in the next few hours.

logan-markewich commented:
Ok, should be good to go I think! One last thing @tmynn the import to ext system info was not working. I hardcoded the default, but feel free to correct me (I just guessed haha)

logan-markewich commented:
Well, I wish I could fully test the notebook. I run into this error... I'll have another look tomorrow, unless someone is able to confirm everything works :) @Disiok @jerryjliu 
![image](https://github.com/jerryjliu/llama_index/assets/22285038/b52fd4f2-1cbc-4efb-801d-5d43a486efd2)

jerryjliu commented on docs/examples/callbacks/AimCallback.ipynb:
```diff
@@ -0,0 +1,150 @@
+            "metadata": {},
+            "outputs": [],
+            "source": [
+                "docs = SimpleDirectoryReader(\"../paul_graham_essay/data\").load_data()"
```
this path needs to be changed to ../../data/paul_graham

jerryjliu commented on docs/examples/callbacks/AimCallback.ipynb:
```diff
@@ -0,0 +1,150 @@
+            "id": "17d1763e",
+            "metadata": {},
+            "source": [
+                "Let's read the documents using `SimpleDirectoryReader` from 'examples/paul_graham_essay/data'."
```
path needs updating (see below)

jerryjliu reviewed (approved):
accepting with one nit on the notebook

jerryjliu reviewed (approved):
accepting with one nit on the notebook
